### ISSUE TITLE
CORs header missing from sign in callback request

### ISSUE THREAD

**Issue by OliviaLB** [authts/oidc-client-ts#1347]
I've set up an auth.service.ts file with this package in a react application. After looking at a few similar issues I've managed to configure my request config and added the necessary metadata that allows my first redirect to succeed to get to the IDS login page (I was initially having CORs errors but adding the metadata resolved it for the initial sign in request and redirects me to our OIDC login page). 

After logging in this then redirects to https://localhost:3000/signin-callback where the secondary part of the PKCE token flow is initiated and the userManager.signinRedirectCallback is fired to then request and retrieve the token to complete authentication. However, this secondary request seems to be missing the necessary CORs request headers for the request to succeed.

I found that when attempting to utilise the chrome extension CORS unblock the request succeeds and token is fetched with the response headers now containing the values for:
- Access-Control-Allow-Credentials: true
- Access-Control-Allow-Methods: *
- Access-Control-Allow-Origin: https://localhost:3000
Whereas when this extension is disabled these headers are missing and therefore the token request fails to complete.

I've tried to pass the above values in through the extraHeaders but the request still seems to be failing. I'd appreciate any help with this as the code is almost functional it is just this issue that seems to be the last obstacle. Below is my current configuration 

```
export default class AuthService implements IAuthService {
    private userManager: UserManager;

    // Constructor to initialise the UserManager with settings
    constructor() {
        const stsSettings: UserManagerSettings = {
            authority: config.OIDCIssuer,
            client_id: config.ClientId,
            redirect_uri: `${config.OIDCRedirectURI}signin-callback`,
            silent_redirect_uri: config.OIDCRedirectURI,
            scope: config.Scope,
            response_type: "code",
            post_logout_redirect_uri: config.OIDCRedirectURI,
            metadata: {
                issuer: config.OIDCIssuer,
                authorization_endpoint: `${config.OIDCIssuer}/connect/authorize`,
                token_endpoint: `${config.OIDCIssuer}/connect/token`,
                userinfo_endpoint: `${config.OIDCIssuer}/connect/userinfo`,
                end_session_endpoint: `${config.OIDCIssuer}/connect/endsession`
            },
            extraTokenParams: {
                "Access-Control-Allow-Credentials": true,
                "Access-Control-Allow-Methods": "*",
                "Access-Control-Allow-Origin": config.OIDCRedirectURI
            }
        };
        this.userManager = new UserManager(stsSettings);

        // Add event listeners for token expiration and silent renew errors
        this.userManager.events.addAccessTokenExpiring(() => {
            console.log("Token is expiring, attempting silent signin...");
            this.signinSilently();
        });

        this.userManager.events.addSilentRenewError((error) => {
            console.log("Silent renew error:", error.message);
        });
    }

    // Initiate the signin process, redirecting the user to the identity provider
    signin(): Promise<void> {
        console.log("trying to redirect");
        return this.userManager.signinRedirect({ state: window.location.href });
    }

    // Complete the signin process, handling the callback after redirection
    completeSignin(): Promise<User> {
        const returnUrl =
            window.location.href !== "/" ? window.location.href : "";
        return this.userManager
            .signinRedirectCallback(returnUrl)
            .then((user) => {
                return user;
            });
    }

// Attempt to signin silently, renewing the access token without user interaction
    signinSilently(): Promise<User | void> {
        console.log("Attempting silent signin...");
        return this.userManager.signinSilent().then(
            (user) => {
                console.log("Silent signin successful");
                return user;
            },
            (error) => {
                console.log("Silent signin failed:", error);
                // If silent signin fails, initiate a regular signin
                this.signin();
            }
        );
    }

    // Handle the silent signin callback
    silentSignInCallback(): Promise<void> {
        return this.userManager.signinSilentCallback() as Promise<void>;
    }

    // Initiate the signout process, redirecting the user to the identity provider
    signout(): Promise<void> {
        return this.userManager.signoutRedirect();
    }

    // Get the access token if available and not expired
    getAccessToken(): Promise<string | null> {
        return this.userManager.getUser().then((user) => {
            if (!!user && !user.expired) {
                return user.access_token;
            } else {
                return null;
            }
        });
    }
}

```

**Comment by pamapa** [comment]
Typically for CORS is that you must properly configure the authorization server, so its knows your host and callback URL...

**Comment by OliviaLB** [comment]
Both permitted CORs origin and redirect URI are in the relevant database, and the fact the first request passes and redirects the user to the ids login page shows that the database recognises the provided parameters. From what I can tell looking at the requests the only thing apparently missing in the secondary callback request is the Access-Control-Allow-Origin: https://localhost:3000 header.

**Comment by OliviaLB** [comment]
![image](https://github.com/authts/oidc-client-ts/assets/99125297/dc2988bf-9df0-438a-aee8-d805e3ec563c)
The missing header also matches the console error  being logged as it specifies Access-Control-Allow-Origin is missing and hence the request is failing.

**Comment by pamapa** [comment]
`Access-Control-Allow-Origin`, .. are part of the response header coming from the authorization server...

**Comment by JMZiem** [comment]
Well that's not true for all OIDC servers and plugins, for example Shibboleth OIDC requires that you send the Access-Control-Allow-Origin "*". If you don't, your browser simply blocks the request - even if everything's fine. An IDP don't has to respond with the correct origin if not requested. And this is exactly the problem, it's not requested!
I have exactly the same problen, extraHeaders like { "Access-Control-Allow-Origin": "*", ...} are simply ignored/cleared before any post request. The response won't have the allow origin header and your browser blocks.

**Comment by zach-betz-hln** [comment]
Hi @OliviaLB and @JMZiem - as was mentioned, the `access-control-allow-origin` header is set by the auth server.

For example, in the Keycloak auth server, you can set this `access-control-allow-origin` header through the public client settings, via the "Web origins" item. Sample screenshot:

<img width="1728" alt="Image" src="https://github.com/user-attachments/assets/edbd13ec-a822-4764-843b-b9224fad1842" />

---

And here's a sample screenshot of the header from the devtools of the [sample](https://github.com/authts/sample-keycloak-react-oidc-context) app:

<img width="1728" alt="Image" src="https://github.com/user-attachments/assets/138b5739-1865-40d2-aa9a-f1e5c86b91e1" />

**Comment by JMZiem** [comment]
Hi @zach-betz-hln, you are of course right about the server and its configuration. However, if the developer does not have access to the server, but only needs to use it - for example, a third-party OIDC server or plugins for servers that map OIDC flows, you need this option. One example is Shibboleth, which provides the flow via an OIDC plugin, but does not send this header by default. I don't understand, why this library is blocking extra headers which are required in several usecases.
So initial the question of @OliviaLB and me was not about the server configuration itself, but an option to use OIDC without CORS problems on (probably misconfigured) third party IDPs and/or plugins.

**Comment by voneiden** [comment]
> Well that's not true for all OIDC servers and plugins, for example Shibboleth OIDC requires that you send the Access-Control-Allow-Origin "*".

@JMZiem  The behaviour you describe sounds very unusual to me. I doubt Shibboleth itself behaves that way, but of course there might be a proxy in between that does weird things. I think the most realistic options here are

a) fork oidc-client-ts
b) deploy your own proxy that has sane CORS behaviour

**Comment by jcphill** [comment]
Forking oidc-client-ts isn't going to solve this problem because the token request is being blocked by the browser based on the same-origin policy before the library gets anywhere near it. There is nothing you can do in JavaScript to fix this. The token endpoint must explicitly trust your origin or the browser will block access from JavaScript to prevent cross-site scripting attacks. See https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy

**Comment by voneiden** [comment]
Indeed, sorry, I missed from JM's first comment that it's the browser that blocks the request. So (a) is out of the window.

**Comment by jcphill** [comment]
Here are the docs for enabling CORS in Shibboleth: https://shibboleth.atlassian.net/wiki/spaces/IDP5/pages/3199501090/CORS+handling

**Comment by JMZiem** [comment]
I'm a little irritated that you don't want to include the simplest solution as an option, even if it's not a security problem or workaround. Shibboleth is unfortunately a very commonly used standard for authentication at universities. The only option requested is to set the origin header in the http request, as no default origin is returned for the "all origins" configuration of Shibboleth server, because (optionally) setting the origin in the request header of the post request to the Auth request completely eliminates the problem.

You might close this ticket as won't do as i solved this problem at our side by not using oidc-client, but for other oidc-client users this information might be interesting.

**Comment by jcphill** [comment]
I don't see any headers other than Content-Type and Authorization being set in the old oidc-client code at https://github.com/DuendeArchive/identity-model-oidc-client-js/blob/edec8f59897bdeedcb0b4167586d49626203c2c1/src/JsonService.js#L99

You also seem to be confusing the Origin request header sent by the browser with the Access-Control-Allow-Origin response header sent by the server.

**Comment by JMZiem** [comment]
Thank you for your patience and clarification. I'm thank you all for your time and effort with this library - i don't want to complain. The initial question was not mine, I just stumbled across the same problem and was a bit unspecific in the course of the discussion. In short, the extraHeaders that can be passed in the requests are ignored in two places as shown above, but a fix might lead to security feature bypassing.
